spec/03-plugins/25-oauth2/03-access_spec.lua

```diff
@@ -2905,6 +2935,76 @@ describe("Plugin: oauth2 [#" .. strategy .. "]", function()
         })
         assert.res_status(401, res)
       end)
+      it("fails to refresh token when scope is mismatching", function ()
+        -- provision code
+        local code, body, res
+        local request_client = helpers.proxy_ssl_client()
+        body = {
+            provision_key = "provision123",
+            client_id = "clientid123",
+            response_type = "code",
+            scope = "scope18",
+            state = "hello",
+            authenticated_userid = "userid123",
+        }
+        res = assert(request_client:send {
+          method = "POST",
+          path = "/oauth2/authorize",
+          body = body,
+          headers = kong.table.merge({
+            ["Host"] = "oauth2_18.com",
+            ["Content-Type"] = "application/json"
+          })
+        })
+        res = assert(cjson.decode(assert.res_status(200, res)))
+        if res.redirect_uri then
+          local iterator, err = ngx.re.gmatch(res.redirect_uri, "^http://google\\.com/kong\\?code=([\\w]{32,32})&state=hello$")
+          assert.is_nil(err)
+          local m, err = iterator()
+          assert.is_nil(err)
+          code = m[1]
+        end
+
+        -- provision token
+        body = {
+          code = code,
+          client_id = "clientid123",
+          client_secret = "secret123",
+          grant_type = "authorization_code",
+          redirect_uri = "http://google.com/kong",
+        }
+        res = assert(request_client:send {
+          method  = "POST",
+          path    = "/oauth2/token",
+          body    = body,
+          headers = {
+            ["Host"]         = "oauth2_18.com",
+            ["Content-Type"] = "application/json"
+          }
+        })
+        local token = assert(cjson.decode(assert.res_status(200, res)))
+
+        -- refresh token with mismatching scope
+        res = assert(request_client:send {
+          method  = "POST",
+          path    = "/oauth2/token",
+          body    = {
+            refresh_token    = token.refresh_token,
+            client_id        = "clientid123",
+            client_secret    = "secret123",
+            grant_type       = "refresh_token",
+          },
+          headers = {
+            ["Host"]         = "oauth2_19.com",
+            ["Content-Type"] = "application/json"
+          }
+        })
+        res = assert(cjson.decode(assert.res_status(400, res)))
+        assert.same({
+          error = "invalid_scope",
+          error_description = "scope mismatch"
+        }, res)
+      end)
```

Comment:
Please add a test case that verifies the subsetting functionality.

Reply:
addressed